packages/vite/src/node/preview.ts

```diff
@@ -129,16 +132,59 @@ export async function preview(
       }
     },
   })
-  app.use(previewBase, async (req, res, next) => {
-    if (shouldServe(req.url!, distDir)) {
-      return assetServer(req, res, next)
+  app.use(
+    previewBase,
+    // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+    async function vitePreviewStaticMiddleware(req, res, next) {
+      if (shouldServe(req.url!, distDir)) {
+        return assetServer(req, res, next)
+      }
+      next()
+    },
+  )
+
+  // html fallback
+  if (config.appType === 'spa' || config.appType === 'mpa') {
+    // append trailing slash when base didn't have it
+    if (config.rawBase !== config.base) {
+      // Keep the named function. The name is visible in debug logs via `DEBUG=connect:dispatcher ...`
+      app.use(function viteRewriteBaseAccessWithoutTrailingSlashMiddleware(
+        req,
+        res,
+        next,
+      ) {
+        try {
+          const pathname = decodeURIComponent(
+            new URL(req.url!, 'http://example.com').pathname,
+          )
+          if (pathname === config.rawBase) {
+            req.url = config.base + req.url!.slice(config.rawBase.length)
```

Comment:
there's a `stripBase` in `utils` that you might be able to use for this

Reply:
This line is replacing `/foo` with `/foo/`. IIUC it's different from what `stripBase` does.